tls/s2n_cipher_suites.c

```diff
@@ -1168,13 +1216,175 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
+            return false;
+        }
+
+        /* Make sure the cipher is valid for available certs */
+        if (s2n_is_cipher_suite_valid_for_auth(conn, potential_match) != S2N_SUCCESS) {
+            return false;
+        }
+
+        /* TLS 1.3 does not include key exchange in cipher suites */
+        if (potential_match->minimum_required_tls_version < S2N_TLS13) {
+            /* If the kex is not supported continue to the next candidate */
+            bool kex_supported = false;
+            POSIX_GUARD_RESULT(s2n_kex_supported(potential_match, conn, &kex_supported));
+            if (!kex_supported) {
+                return false;
+            }
+
+            /* If the kex is not configured correctly continue to the next candidate */
+            if (s2n_result_is_error(s2n_configure_kex(potential_match, conn))) {
+                return false;
+            }
+        }
+
+        /**
+        *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
+        *# The server MUST ensure that it selects a compatible PSK
+        *# (if any) and cipher suite.
+        **/
+        if (conn->psk_params.chosen_psk != NULL) {
+            if (potential_match->prf_alg != conn->psk_params.chosen_psk->hmac_alg) {
+                return false;
+            }
+        }
+
+        return true;
+}
+
+/**
+ * @brief Same as s2n_wire_ciphers_contain() except that it returns the match's index in the client wire.
+ * 
+ * @param match Cipher suite IANA which we wish to find a match for
+ * @param wire Client wire
+ * @param count Number of entries
+ * @param cipher_suite_len Length offset
+ * @return int Index of the match in the client wire. If a match was not found then -1 is returned.
+ */
+static int s2n_wire_ciphers_has_server_cipher_at(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
+{
+    for (uint32_t i = 0; i < count; i++) {
+        const uint8_t *theirs = wire + (i * cipher_suite_len) + (cipher_suite_len - S2N_TLS_CIPHER_SUITE_LEN);
+
+        if (!memcmp(match, theirs, S2N_TLS_CIPHER_SUITE_LEN)) {
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+/**
+ * @brief This is the logic for parsing a list of s2n_cipher_suites* and selecting a cipher_suite accounting
+ *        for equal preference grouping. This method assumes that the server side cipher suite preference is
+ *        correctly formatted eg for every group_start delimiter, there exists a following group_end delimiter.
+ *        TODO: Add reference to documentation and add basic validation
+ * 
+ * @param conn s2n_connection is used to ensure the negotiated cipher is valid for the connection
+ * @param wire client wire containing cipher suite IANAs
+ * @param count Number of entries
+ * @param cipher_suite_len Cipher suite length offset
+ * @return int Server's index of the negotiated cipher suite. If no cipher suite was selected then -1 is returned.
+ */
+static int s2n_get_negotiated_server_index(struct s2n_connection* conn,
```

Comment:
Again, doxygen comments aren't needed here, and this should return a S2N_RESULT.